search/materialize.go

```diff
@@ -0,0 +1,420 @@
+// Copyright The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package search
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"runtime"
+	"slices"
+	"sort"
+
+	"github.com/efficientgo/core/errors"
+	"github.com/parquet-go/parquet-go"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/prometheus-community/parquet-common/schema"
+	"github.com/prometheus-community/parquet-common/util"
+)
+
+type Materializer struct {
+	lf *parquet.File
+	cf *parquet.File
+	s  *schema.TSDBSchema
+	d  *schema.PrometheusParquetChunksDecoder
+
+	colIdx      int
+	concurrency int
+
+	dataColToIndex []int
+}
+
+func NewMaterializer(s *schema.TSDBSchema, d *schema.PrometheusParquetChunksDecoder, lf, cf *parquet.File) (*Materializer, error) {
```

Comment:
This is a minor nit, but it makes no sense to me that the Materializer is public if its only method is the "Materialize" function ( which takes unexported slice of rowRanges ); wouldnt it be better to make that an implementation detail of the package and just call "Materialize(....)" which then creates a private materializer and does its thing?

Reply:
can we make the interface just `Materialize(schema, decoder, file1, file2) ([]storage.ChunkSeries, error)` ?

Reply:
`series, err := search.Materialize(s, d, lf, cf, constraints...) ` feels right to me as interface for the package

Reply:
I like this because down the line if we want to try a different approach for "materializing" we can simply implement a new materializer and benchmark against it.

Reply:
It feels like this package leaks implementation details - I would prefer having a proper interface to use; if we expose everything then clients start depending on everythin and we have hard time changing stuff from the get go, i think thats a mistake. Lets only export the bare minimum

Reply:
imho interface to this package should just be ```
func Match(
	ctx context.Context,
	labelPfile *parquet.File,
	labelSchema *parquet.Schema,
	chunkPfile *parquet.File,
	chunkSchema *parquet.Schema,
	cs ...Constraint,
) ([]SomeSeriesThing, error) {

```

Reply:
U wanna force the user to use all or nothing? I dont think we should enforce that!

I can see projects like cortex maybe want to have a slightly different implementations on some places but still use Materialize for instance.

Reply:
Lets do it like this: for now we export, eventually we take stuff as interfaces ( once we have ironed out what is useful etc )